src/PointwiseFunctions/Hydro/EquationsOfState/Tabulated3d.cpp

```diff
@@ -0,0 +1,436 @@
+      electron_fraction, rest_mass_density, temperature);
+
+  Scalar<DataType> cs2 =
+      make_with_value<Scalar<DataType>>(get(rest_mass_density), 0.0);
+
+  if constexpr (std::is_same_v<DataType, double>) {
+    auto weights = interpolator_.get_weights(get(log_temperature),
+                                             get(log_rest_mass_density),
+                                             get(converted_electron_fraction));
+    auto interpolated_state =
+        interpolator_.template interpolate<CsSquared>(weights);
+    get(cs2) = interpolated_state[0];
+
+  } else if constexpr (std::is_same_v<DataType, DataVector>) {
+    for (size_t s = 0; s < electron_fraction.size(); ++s) {
+      auto weights = interpolator_.get_weights(
+          get(log_temperature)[s], get(log_rest_mass_density)[s],
+          get(converted_electron_fraction)[s]);
+      auto interpolated_state =
+          interpolator_.template interpolate<CsSquared>(weights);
+      get(cs2)[s] = interpolated_state[0];
+    }
+  }
+
+  return cs2;
+}
+
+template <bool IsRelativistic>
+double Tabulated3D<IsRelativistic>::specific_internal_energy_lower_bound(
+    const double rest_mass_density, const double electron_fraction) const {
+  double converted_electron_fraction =
+      std::min(std::max(electron_fraction_lower_bound(), electron_fraction),
+               electron_fraction_upper_bound());
+
+  double log_rest_mass_density =
+      std::min(std::max(rest_mass_density_lower_bound(), rest_mass_density),
+               rest_mass_density_upper_bound());
+
+  log_rest_mass_density = log(log_rest_mass_density);
+
+  auto weights = interpolator_.get_weights(log(temperature_lower_bound()),
+                                           log_rest_mass_density,
+                                           converted_electron_fraction);
+  auto interpolated_state =
+      interpolator_.template interpolate<Epsilon>(weights);
+
+  return exp(interpolated_state[0]) + energy_shift_;
+}
+
+template <bool IsRelativistic>
+double Tabulated3D<IsRelativistic>::specific_internal_energy_upper_bound(
+    const double rest_mass_density, const double electron_fraction) const {
+  double converted_electron_fraction =
+      std::min(std::max(electron_fraction_lower_bound(), electron_fraction),
+               electron_fraction_upper_bound());
+
+  double log_rest_mass_density =
+      std::min(std::max(rest_mass_density_lower_bound(), rest_mass_density),
+               rest_mass_density_upper_bound());
+
+  log_rest_mass_density = log(log_rest_mass_density);
+
+  auto weights = interpolator_.get_weights(
+      log(upper_bound_tolerance_ * temperature_upper_bound()),
+      log_rest_mass_density, converted_electron_fraction);
+  auto interpolated_state =
+      interpolator_.template interpolate<Epsilon>(weights);
+
+  return exp(interpolated_state[0]) + energy_shift_;
+}
+
+template <bool IsRelativistic>
+Tabulated3D<IsRelativistic>::Tabulated3D(std::vector<double> electron_fraction,
```

Comment:
Okay, clang-tidy is unhappy because it doesn't seem to be able to figure out that no extra copies are done here :shrug: I was hoping/expecting it to be able to analyze this correctly.

This is how I would get clang-tidy to be happy (i.e. tell it that it's wrong 🙃 ):
```cpp
Tabulated3D<IsRelativistic>::Tabulated3D(// NOLINTNEXTLINE(performance-unnecessary-value-param)
                                         std::vector<double> electron_fraction,
                                         // NOLINTNEXTLINE(performance-unnecessary-value-param)
                                         std::vector<double> log_density,
                                         // NOLINTNEXTLINE(performance-unnecessary-value-param)
                                         std::vector<double> log_temperature,
                                         // NOLINTNEXTLINE(performance-unnecessary-value-param)
                                         std::vector<double> table_data,
                                         double energy_shift,
                                         double enthalpy_minimum) {
```
You can, of course, squash directly again. All other clang-tidy complaints are about boost internals that we have no control over and so can just ignore.

Reply:
Well, would have been too easy if clang-tidy got it the first time ^^
Ill added those pragmas, and squashed. Thanks!

Reply:
Always :D The dream is where things Just Work ™️